File: obsolete/android/app/src/main/java/de/familienkalender/app/ui/meals/AiMealPlanViewModel.kt
package de.familienkalender.app.ui.meals

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import de.familienkalender.app.data.remote.dto.*
import de.familienkalender.app.data.repository.AiRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.format.DateTimeFormatter

enum class AiStep { LOADING, CONFIG, GENERATING, PREVIEW, ERROR }

data class AiMealPlanUiState(
    val step: AiStep = AiStep.LOADING,
    val weekStart: String = "",
    val availableDates: List<String> = emptyList(),
    val occupiedSlots: Set<String> = emptySet(),
    val selectedSlots: Set<String> = emptySet(),
    val cookidooAvailable: Boolean = false,
    val includeCookidoo: Boolean = false,
    val servings: Int = 4,
    val preferences: String = "",
    val preview: PreviewMealPlanResponse? = null,
    val isConfirming: Boolean = false,
    val error: String? = null
)

class AiMealPlanViewModel(
    private val aiRepository: AiRepository,
    private val weekStart: String
) : ViewModel() {

    private val _uiState = MutableStateFlow(AiMealPlanUiState(weekStart = weekStart))
    val uiState: StateFlow<AiMealPlanUiState> = _uiState

    init {
        loadAvailableRecipes()
    }

    private fun loadAvailableRecipes() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(step = AiStep.LOADING)
            aiRepository.getAvailableRecipes(weekStart).fold(
                onSuccess = { response ->
                    val monday = LocalDate.parse(weekStart)
                    val dates = (0..6).map { monday.plusDays(it.toLong()).format(DateTimeFormatter.ISO_DATE) }
                    val occupied = mutableSetOf<String>()
                    response.currentSlots.forEach { (date, slots) ->
                        slots.forEach { (slot, info) ->
                            if (info != null) occupied.add("$date|$slot")
                        }
                    }
                    val emptySlots = dates.flatMap { d ->
                        listOf("lunch", "dinner").mapNotNull { s ->
                            val id = "$d|$s"
                            if (!occupied.contains(id)) id else null
                        }
                    }.toSet()

                    _uiState.value = _uiState.value.copy(
                        step = AiStep.CONFIG,
                        availableDates = dates,
                        occupiedSlots = occupied,
                        selectedSlots = emptySlots,
                        cookidooAvailable = response.cookidooAvailable
                    )
                },
                onFailure = { e ->
                    _uiState.value = _uiState.value.copy(step = AiStep.ERROR, error = e.message)
                }
            )
        }
    }

    fun toggleSlot(slotId: String) {
        val current = _uiState.value.selectedSlots.toMutableSet()
        if (current.contains(slotId)) current.remove(slotId) else current.add(slotId)
        _uiState.value = _uiState.value.copy(selectedSlots = current)
    }

    fun setIncludeCookidoo(v: Boolean) { _uiState.value = _uiState.value.copy(includeCookidoo = v) }
    fun setServings(v: Int) { _uiState.value = _uiState.value.copy(servings = v) }
    fun setPreferences(v: String) { _uiState.value = _uiState.value.copy(preferences = v) }

    fun generate() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(step = AiStep.GENERATING)
            val slots = _uiState.value.selectedSlots.map { id ->
                val parts = id.split("|")
                SlotSelection(parts[0], parts[1])
            }
            val request = GenerateMealPlanRequest(
                weekStart = weekStart,
                servings = _uiState.value.servings,
                preferences = _uiState.value.preferences,
                selectedSlots = slots,
                includeCookidoo = _uiState.value.includeCookidoo
            )
            aiRepository.generateMealPlan(request).fold(
                onSuccess = { preview ->
                    _uiState.value = _uiState.value.copy(step = AiStep.PREVIEW, preview = preview)
                },
                onFailure = { e ->
                    _uiState.value = _uiState.value.copy(step = AiStep.ERROR, error = e.message)
                }
            )
        }
    }

    fun confirm(onConfirmed: (List<Int>) -> Unit) {
        viewModelScope.launch {
            val preview = _uiState.value.preview ?: return@launch
            _uiState.value = _uiState.value.copy(isConfirming = true)
            val request = ConfirmMealPlanRequest(weekStart = weekStart, items = preview.suggestions)
            aiRepository.confirmMealPlan(request).fold(
                onSuccess = { response ->
                    _uiState.value = _uiState.value.copy(isConfirming = false)
                    onConfirmed(response.mealIds)
                },
                onFailure = { e ->
                    _uiState.value = _uiState.value.copy(isConfirming = false, step = AiStep.ERROR, error = e.message)
                }
            )
        }
    }

    fun backToConfig() {
        _uiState.value = _uiState.value.copy(step = AiStep.CONFIG, error = null)
    }

    class Factory(
        private val aiRepository: AiRepository,
        private val weekStart: String
    ) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return AiMealPlanViewModel(aiRepository, weekStart) as T
        }
    }
}

File: obsolete/ios/Familienkalender/ViewModels/AIMealPlanViewModel.swift
import Foundation

@Observable
@MainActor
final class AIMealPlanViewModel {

    // MARK: - Wizard Steps

    enum WizardStep {
        case config, loading, preview, confirming, done
    }

    // MARK: - State

    var availableRecipes: AvailableRecipesResponse?
    var selectedSlots: Set<String> = []
    var servings: Int = 4
    var preferences: String = ""
    var includeCookidoo: Bool = false
    var preview: PreviewMealPlanResponse?
    var confirmResult: ConfirmMealPlanResponse?
    var step: WizardStep = .config
    var isLoading: Bool = false
    var errorMessage: String?

    // MARK: - Dependencies

    private let aiRepo: AIRepository

    init(aiRepo: AIRepository) {
        self.aiRepo = aiRepo
    }

    // MARK: - Slot key helpers

    private func slotKey(date: String, slot: String) -> String {
        "\(date)_\(slot)"
    }

    // MARK: - Load

    func loadAvailableRecipes(weekStart: String) async {
        isLoading = true
        errorMessage = nil
        do {
            let response = try await aiRepo.getAvailableRecipes(weekStart: weekStart)
            availableRecipes = response

            selectedSlots.removeAll()
            for emptySlot in response.emptySlots {
                selectedSlots.insert(slotKey(date: emptySlot.date, slot: emptySlot.slot))
            }
        } catch {
            errorMessage = "Verfuegbare Rezepte konnten nicht geladen werden: \(error.localizedDescription)"
        }
        isLoading = false
    }

    // MARK: - Generate

    func generatePlan(weekStart: String) async {
        step = .loading
        errorMessage = nil

        let slots: [SlotSelection] = selectedSlots.compactMap { key in
            let parts = key.split(separator: "_", maxSplits: 1)
            guard parts.count == 2 else { return nil }
            return SlotSelection(date: String(parts[0]), slot: String(parts[1]))
        }

        let request = GenerateMealPlanRequest(
            weekStart: weekStart,
            servings: servings,
            preferences: preferences,
            selectedSlots: slots,
            includeCookidoo: includeCookidoo
        )

        do {
            preview = try await aiRepo.generateMealPlan(request)
            step = .preview
        } catch {
            errorMessage = "KI-Plan konnte nicht generiert werden: \(error.localizedDescription)"
            step = .config
        }
    }

    // MARK: - Confirm

    func confirmPlan(weekStart: String) async -> ConfirmMealPlanResponse? {
        guard let preview = preview else { return nil }
        step = .confirming
        errorMessage = nil

        let request = ConfirmMealPlanRequest(
            weekStart: weekStart,
            items: preview.suggestions
        )

        do {
            let result = try await aiRepo.confirmMealPlan(request)
            confirmResult = result
            step = .done
            return result
        } catch {
            errorMessage = "Plan konnte nicht bestaetigt werden: \(error.localizedDescription)"
            step = .preview
            return nil
        }
    }

    // MARK: - Reset

    func reset() {
        step = .config
        preview = nil
        confirmResult = nil
        errorMessage = nil
        preferences = ""
    }
}
